sys/include/net/gnrc/ipv6/nib.h

```diff
@@ -0,0 +1,541 @@
+    GNRC_IPV6_NIB_NUD_STATE_STALE,          /**< address might not be reachable */
+    GNRC_IPV6_NIB_NUD_STATE_DELAY,          /**< NUD will be performed in a moment */
+    GNRC_IPV6_NIB_NUD_STATE_PROBE,          /**< NUD is performed */
+    GNRC_IPV6_NIB_NUD_STATE_REACHABLE,      /**< entry is reachable */
+} gnrc_ipv6_nib_nud_state_t;
+
+/**
+ * @brief States for 6LoWPAN address registration (6Lo-AR)
+ */
+typedef enum {
+    GNRC_IPV6_NIB_AR_STATE_NONE = 0,        /**< not managed by 6Lo-AR */
+    GNRC_IPV6_NIB_AR_STATE_GC,              /**< address can be removed when low on memory */
+    GNRC_IPV6_NIB_AR_STATE_TENTATIVE,       /**< address registration still pending */
+    GNRC_IPV6_NIB_AR_STATE_REGISTERED,      /**< address is registered */
+} gnrc_ipv6_nib_ar_state_t;
+
+/**
+ * @brief   A public NIB entry
+ */
+typedef struct {
+    const ipv6_addr_t dst;          /**< A destination or prefix to destination */
+    const ipv6_addr_t next_hop;     /**< Next hop to gnrc_ipv6_nib_t::dst */
+#if defined(MODULE_GNRC_SIXLOWPAN_ND_ROUTER) || defined(DOXYGEN)
+    /**
+     * @brief   The neighbors EUI-64 (used for DAD)
+     *
+     * 
+     */
+    eui64_t eui64;
+#endif
+    const uint8_t l2addr[GNRC_NETIF_HDR_L2ADDR_MAX_LEN];
+                                    /**< L2 address of next hop */
+    const unsigned pfx_len : 8;     /**< prefix-length of gnrc_ipv6_nib_t::dst */
+    const unsigned l2addr_len : 4;  /**< length of gnrc_ipv6_nib_t::l2addr */
+    const gnrc_ipv6_nib_nud_state_t nud_state : 3;
+                                    /**< NUD state of gnrc_ipv6_nib_t::next_hop */
+    const bool is_router : 1;       /**< gnrc_ipv6_nib_t::next_hop is router */
+    const unsigned iface : 5;       /**< interface to gnrc_ipv6_nib_t::dst */
+    const gnrc_ipv6_nib_ar_state_t ar_state : 2;
+                                    /**< 6Lo-AR state of gnrc_ipv6_nib_t::next_hop */
+    /**
+     * @brief   Use gnrc_ipv6_nib_t::cid for compression of gnrc_ipv6_nib_t::dst
+     *          when using @ref net_gnrc_sixlowpan_iphc
+     */
+    const bool use_for_comp : 1;
+
+    /**
+     * @brief   Context identifier for gnrc_ipv6_nib_t::dst for stateful header
+     *          compression using @ref net_gnrc_sixlowpan_iphc
+     */
+    const unsigned cid : 4;
+} gnrc_ipv6_nib_t;
+
+/**
+ * @brief   Gets next reachable hop to @p dst
+ *
+ * @param[in] iface Interface to restrict search to search for next reachable
+ *                  hop to. May be KERNEL_PID_UNDEF for any interface.
+ * @param[in] dst   The destination address of @p pkt.
+ * @param[in] pkt   The Packet
+ *
+ * @return The NIB to the reachable entry on success.
+ * @return NULL, if no reachable next hop was found.
+ */
+const gnrc_ipv6_nib_t *gnrc_ipv6_nib_next_hop(unsigned iface,
+                                              const ipv6_addr_t *dst,
+                                              const gnrc_pktsnip_t *pkt);
```

Comment:
From [RFC 4861, 7.2.2](https://tools.ietf.org/html/rfc4861#section-7.2.2):

>   While waiting for address resolution to complete, the sender MUST,
   for each neighbor, retain a small queue of packets waiting for
   address resolution to complete.  The queue MUST hold at least one
   packet, and MAY contain more.  However, the number of queued packets
   per neighbor SHOULD be limited to some small value.  When a queue
   overflows, the new arrival SHOULD replace the oldest entry.  Once
   address resolution completes, the node transmits any queued packets.

To allow for this queuing action the packet must be provided, too (mental note to self: document that).